### PR TITLE
Add nullable annotations to Http.Abstractions, Http.Features, Connections.Abstractions

### PR DESCRIPTION
**Pull request by JamesNK**
Addresses https://github.com/aspnet/AspNetCore/issues/5680

Annotations on projects at the bottom of the dependency tree. Because I haven't converted the tests there may be some types that are not null but should be. Should discover those as more projects are annotated.

### CODE REVIEW

File: src/Http/Headers/src/BaseHeaderParser.cs [hunk 1]
```diff
@@ -14,9 +15,9 @@ protected BaseHeaderParser(bool supportsMultipleValues)
 
         protected abstract int GetParsedValueLength(StringSegment value, int startIndex, out T parsedValue);
 
-        public sealed override bool TryParseValue(StringSegment value, ref int index, out T parsedValue)
+        public sealed override bool TryParseValue(StringSegment value, ref int index, [MaybeNullWhen(false)]out T parsedValue)
```
Comment (halter73) [inline src/Http/Headers/src/BaseHeaderParser.cs]: It looks like TryParseValue can return true even in cases where parsedValue is never set if SupportsMultipleValues is also true.

File: src/Http/Headers/src/HttpHeaderParser.cs [hunk 2]
```diff
@@ -27,7 +28,7 @@ public bool SupportsMultipleValues
         // pointing to the next non-whitespace character after a delimiter. E.g. if called with a start index of 0
         // for string "value , second_value", then after the call completes, 'index' must point to 's', i.e. the first
         // non-whitespace after the separator ','.
-        public abstract bool TryParseValue(StringSegment value, ref int index, out T parsedValue);
+        public abstract bool TryParseValue(StringSegment value, ref int index, [NotNullWhen(true)]out T parsedValue);
```
Comment (halter73) [inline src/Http/Headers/src/HttpHeaderParser.cs]: Shouldn't the type be marked nullable when using attributes like NotNullWhen?

**Review by Tratcher — CHANGES_REQUESTED** [review]
How do you plan to handle microsoft.extensions.activatorutilities.sources?

**Comment by rynowak** [comment]
API review YO

**Comment by anurse** [comment]
Sounds like there are some build issues with using nullable annotations here. @JamesNK you can check in with @dotnet/aspnet-build and see if we can figure this out?

**Comment by dougbu** [comment]
>  Sounds like there are some build issues with using nullable annotations here.

@anurse something beyond the missing `Contract.Requires(...)` calls that @Tratcher mentioned? Could you be a bit more specific?

**Comment by anurse** [comment]
Just what @JamesNK said in Triage ;).

**Comment by JamesNK** [comment]
No immediate plans to look at this. Can re-open when I have time in the future.

**Pull request closed without merging**